### PR TITLE
OHRM5X-2127: Add API to get Menu Items for Mobile

### PR DESCRIPTION
**Pull request by ManulMax**

### CODE REVIEW

File: src/composer.json [hunk 1]
```diff
@@ -117,7 +119,9 @@
             "OrangeHRM\\Tests\\Dashboard\\": "./plugins/orangehrmDashboardPlugin/test",
             "OrangeHRM\\Tests\\LDAP\\": "./plugins/orangehrmLDAPAuthenticationPlugin/test",
             "OrangeHRM\\Tests\\Buzz\\": "./plugins/orangehrmBuzzPlugin/test",
-            "OrangeHRM\\Tests\\Help\\": "./plugins/orangehrmHelpPlugin/test"
+            "OrangeHRM\\Tests\\Claim\\": "./plugins/orangehrmClaimPlugin/test",
```
Comment (RajithaKumara) [inline src/composer.json:122]:
```suggestion
```
remove this line

File: src/composer.json [hunk 2]
```diff
@@ -78,6 +78,8 @@
             "OrangeHRM\\CorporateBranding\\": "./plugins/orangehrmCorporateBrandingPlugin",
             "OrangeHRM\\LDAP\\": "./plugins/orangehrmLDAPAuthenticationPlugin",
             "OrangeHRM\\SystemCheck\\": "./plugins/orangehrmSystemCheckPlugin",
+            "OrangeHRM\\Claim\\": "./plugins/orangehrmClaimPlugin",
```
Comment (RajithaKumara) [inline src/composer.json:81]:
```suggestion
```
remove this line, as claim moved to 5.5 branch

File: src/plugins/orangehrmAttendancePlugin/Api/EmployeeAttendanceSummaryAPI.php [hunk 3]
```diff
@@ -35,15 +35,19 @@
...
 use OrangeHRM\Entity\Employee;
 
 class EmployeeAttendanceSummaryAPI extends Endpoint implements CollectionEndpoint
 {
     use UserRoleManagerTrait;
     use AttendanceServiceTrait;
+    use DateTimeHelperTrait;
 
     public const PARAMETER_DATE = 'date';
+    public const  PARAMETER_FROM_DATE = 'fromDate';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAttendancePlugin/Api/EmployeeAttendanceSummaryAPI.php:49]:
```suggestion
    public const PARAMETER_FROM_DATE = 'fromDate';
```
:relaxed:

File: src/plugins/orangehrmAttendancePlugin/Api/EmployeeAttendanceSummaryAPI.php [hunk 4]
```diff
@@ -57,21 +61,44 @@ public function getAll(): EndpointResult
...
+            RequestParams::PARAM_TYPE_QUERY,
+            self::PARAMETER_FROM_DATE
+        );
+
+        $toDate = $this->getRequestParams()->getStringOrNull(
+            RequestParams::PARAM_TYPE_QUERY,
+            self::PARAMETER_TO_DATE
+        );
+
+        if ($fromDate != null && $toDate != null && strtotime($fromDate) > strtotime($toDate)) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAttendancePlugin/Api/EmployeeAttendanceSummaryAPI.php:79]: *

File: src/plugins/orangehrmAttendancePlugin/test/fixtures/testcases/EmployeeAttendanceSummaryAPITestCases.yaml [hunk 5]
```diff
@@ -248,9 +248,376 @@ GetAll:
...
+      pim.employee_service: \OrangeHRM\Pim\Service\EmployeeService
+      core.authorization.user_role_manager_helper: \OrangeHRM\Core\Authorization\Helper\UserRoleManagerHelper
+    factories:
+      core.authorization.user_role_manager: [ '\OrangeHRM\Core\Authorization\Manager\UserRoleManagerFactory', 'getNewUserRoleManager' ]
+    query:
+      fromDate: '2023-02-10'
+      toDate: '2022-02-14'
+    exception:
+      class: '\OrangeHRM\Core\Api\V2\Exception\InvalidParamException'
+      message: 'Invalid Parameter'
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAttendancePlugin/test/fixtures/testcases/EmployeeAttendanceSummaryAPITestCases.yaml:623]: better to validate using `invalidOnly` attribute

File: src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            RequestParams::PARAM_TYPE_QUERY,
+            self::PARAMETER_FROM_DATE
+        );
+
+        $toDate = $this->getRequestParams()->getDateTimeOrNull(
+            RequestParams::PARAM_TYPE_QUERY,
+            self::PARAMETER_TO_DATE
+        );
+
+        // TODO:: filter by status
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php:67]: TODO

File: src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+        if ($fromDate != null && $toDate != null && $fromDate > $toDate) {
+//            throw $this->getInvalidParamException(["fromDate","toDate"]);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php:70]: TODO

File: src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                    new Rule(Rules::API_DATE)
+                ),
+            ),
+            $this->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::PARAMETER_TO_DATE,
+                    new Rule(Rules::API_DATE)
+                ),
+            ),
+            ...$this->getSortingAndPaginationParamsRules(LeaveRequestSearchFilterParams::ALLOWED_SORT_FIELDS) // TODO:: replace with correct class
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php:114]: TODO

File: src/plugins/orangehrmLeavePlugin/Dao/LeaveRequestDao.php [hunk 9]
```diff
@@ -724,4 +724,70 @@ public function getLeavesByLeaveIds(array $leaveIds): array
...
+    }
+
+    /**
+     * @param int $leaveRequestId
+     * @param string $fromDate
+     * @param string $toDate
+     *
+     * @return array
+     */
+    public function getLeavesFilterByDateRange(
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Dao/LeaveRequestDao.php:769]: do we need this?

File: src/plugins/orangehrmLeavePlugin/config/routes.yaml [hunk 10]
```diff
@@ -207,6 +207,13 @@ apiv2_leave_reports_data:
   defaults:
     _api: OrangeHRM\Leave\Api\LeaveReportDataAPI
 
+apiv2_leave_employee_leave:
+  path: /api/v2/leave/employees/{empNumber}/leaves
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/config/routes.yaml:211]:
```suggestion
  path: /api/v2/leave/employees/leaves
```
:thinking: will go like this, we can utilize for my attendance also

File: src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    public const  PARAMETER_FROM_DATE = 'fromDate';
+    public const PARAMETER_TO_DATE = 'toDate';
+
+    /**
+     * @inheritDoc
+     */
+    public function getAll(): EndpointResult
+    {
+        $employeeNumber = $this->getRequestParams()->getInt(
+            RequestParams::PARAM_TYPE_ATTRIBUTE,
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Api/EmployeeLeaveAPI.php:52]:
```suggestion
            RequestParams::PARAM_TYPE_QUERY,
```

File: src/plugins/orangehrmLeavePlugin/test/Api/EmployeeLeaveAPITest.php [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Api;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/test/Api/EmployeeLeaveAPITest.php:3]:
```suggestion
namespace OrangeHRM\Tests\Leave\Api;
```

File: src/plugins/orangehrmMobilePlugin/Api/MenuItemAPI.php [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    {
+        if (!$this->timesheetPeriodService instanceof TimesheetPeriodService) {
+            $this->timesheetPeriodService = new TimesheetPeriodService();
+        }
+        return $this->timesheetPeriodService->isTimesheetPeriodDefined();
+    }
+
+    /**
+     * @OA\Get(
+     *     path="/api/v2/mobile/modules",
```
Comment (RajithaKumara) [inline src/plugins/orangehrmMobilePlugin/Api/MenuItemAPI.php:147]: change URL

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmMobilePlugin/test/fixtures/testcases/MenuItemAPITestCases.yaml [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+          name: 'Leave List',
+          name: 'Assign Leave'
+        ]
+      1:
+        name: 'Time'
+        children: [
+          name: 'Punch In/Out',
+          name: 'My Attendance',
+          name: 'Employee Attendance'
+        ]
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply src/plugins/orangehrmMobilePlugin/test/fixtures/testcases/MenuItemAPITestCases.yaml:154]: sure

File: src/plugins/orangehrmMobilePlugin/test/Api/MenuItemAPITest.php [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+    public function dataProviderForTestGetOne2(): array
+    {
+        return $this->getTestCases('MenuItemAPITestCases.yaml', 'GetOne - Leave period and Timesheet period not defined');
+    }
+
+    /**
+     * @dataProvider dataProviderForTestGetOne3
+     */
+    public function testGetOne3(TestCaseParams $testCaseParams): void
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply src/plugins/orangehrmMobilePlugin/test/Api/MenuItemAPITest.php:72]: added to PR - https://github.com/orangehrm/orangehrm/pull/1669

File: src/plugins/orangehrmMobilePlugin/test/fixtures/testcases/MenuItemAPITestCases.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (ManulMax) [reply src/plugins/orangehrmMobilePlugin/test/fixtures/testcases/MenuItemAPITestCases.yaml:154]: added to PR - https://github.com/orangehrm/orangehrm/pull/1669

**Pull request merged**